cedar-lean/Cedar/Thm/Partial/Evaluation/WellFormed.lean

```diff
@@ -28,51 +28,53 @@ import Cedar.Thm.Data.Set
 
 namespace Cedar.Spec
 
-/--
-  We define `WellFormed` for `Value` in the obvious way
--/
-def Value.WellFormed (v : Value) : Prop :=
-  match v with
-  | .set s => s.WellFormed
-  | .record r => r.WellFormed
+open Cedar.Data
+
+/-- All `Prim`s are structurally WellFormed. -/
+def Prim.WellFormed : Spec.Prim → Prop
   | _ => true
 
-/--
-  `Request`s are `WellFormed` if the context is `WellFormed`
--/
-def Request.WellFormed (req : Request) : Prop :=
-  req.context.WellFormed
+/-- All `Ext`s are structurally WellFormed. -/
+def Ext.WellFormed : Ext → Prop
+  | .decimal _ => true
+  | .ipaddr _  => true
+
+def Value.WellFormed : Spec.Value → Prop
+  | .prim p => p.WellFormed
+  | .set s => s.WellFormed ∧ ∀ elt ∈ s, elt.WellFormed
+  | .record r => r.WellFormed ∧ ∀ kv ∈ r.kvs, kv.snd.WellFormed
+  | .ext x => x.WellFormed
+decreasing_by
+  all_goals simp_wf
+  case _ h₁ => -- set
+    have := Set.sizeOf_lt_of_mem h₁
+    omega
+  case _ h₁ => -- record
+    have h₂ := Map.sizeOf_lt_of_value h₁
+    apply Nat.lt_trans h₂
+    have h₃ := Map.sizeOf_lt_of_kvs r
+    simp [Map.kvs] at *
+    omega
+
+def Request.WellFormed (req : Spec.Request) : Prop :=
+  req.context.WellFormed ∧ ∀ val ∈ req.context.values, val.WellFormed
 
 end Cedar.Spec
 
 namespace Cedar.Partial
 
-/--
-  We define `WellFormed` for `Partial.Value` using `Spec.Value.WellFormed`
--/
-def Value.WellFormed (pval : Partial.Value) : Prop :=
-  match pval with
+def Value.WellFormed : Partial.Value → Prop
   | .value v => v.WellFormed
   | .residual _ => true
 
-/--
-  `Partial.Request`s are `AllWellFormed` if the context is `WellFormed` and
-  all the context's constituent `Partial.RestrictedValue`s are also `WellFormed`.
-  (principal, action, and resource are always well-formed)
--/
-def Request.AllWellFormed (preq : Partial.Request) : Prop :=
-  preq.context.WellFormed ∧ ∀ rpval ∈ preq.context.values, rpval.WellFormed
+def Request.WellFormed (preq : Partial.Request) : Prop :=
+  preq.context.WellFormed ∧ ∀ pval ∈ preq.context.values, pval.WellFormed
 
-/--
-  We define `WellFormed` for `Partial.EntityData` in the obvious way
--/
 def EntityData.WellFormed (edata : Partial.EntityData) : Prop :=
-  edata.attrs.WellFormed ∧ edata.ancestors.WellFormed
+  edata.attrs.WellFormed ∧
+  edata.ancestors.WellFormed ∧
+  ∀ pval ∈ edata.attrs.values, pval.WellFormed
 
-/--
-  `Partial.Entities` are `AllWellFormed` if they are `WellFormed` and all the
-  constituent `Partial.EntityData` are also `WellFormed`
--/
 def Entities.AllWellFormed (entities : Partial.Entities) : Prop :=
   entities.WellFormed ∧ ∀ edata ∈ entities.values, edata.WellFormed
```

Comment:
Nit: it would feel more consistent if this predicate was called `WellFormed` instead of `AllWellFormed`. I guess this would require changes the definition of `PartialEntities` -- would that break everything?

Reply:
Yeah, since `Partial.Entities` is just an alias for a particular `Map` type, `Partial.Entities.WellFormed` is already defined and means just that the `Map` is `Map.WellFormed`.  If we made it a struct and not an alias we could reclaim the `WellFormed` name, and probably we should

Reply:
Did the replacement, no more `AllWellFormed`